### PR TITLE
CSHARP-2269: Fix Select query when there are Linq query members from different expression levels.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**
For now, I didn't check the possible solution via `Original` field in `FieldExpression`. I will do it tomorrow morning maybe with a small refactoring.

The current solution covers the only case (with `Select` method) which has been described in the ticket.
But I guess similar behavior should be added to some other `Translate` methods, for example to the `TranslateField`. I made a draft commit with these changes (with a question which is currently not clear for me): https://github.com/DmitryLukyanov/mongo-csharp-driver/commit/c24d318c78b3ff10f4e205224491846cdbac49a7.

### CODE REVIEW

File: tests/MongoDB.Driver.Tests/Linq/MongoQueryableTests.cs [hunk 1]
```diff
@@ -1136,6 +1133,33 @@ public void Select_followed_by_group()
...
+                query,
+                2,
+                "{ '$project' : { 'Result' : { '$filter' : { 'input' : '$G', 'as' : 'g1', 'cond' : { '$gt' : ['$$g1.N', { '$min' : '$G.N' }] } } }, '_id' : 0 } }");
+
+            commonResult[0].Result.Count().Should().Be(1);
+            commonResult[0].Result.First().N.Should().Be(2);
+            commonResult[1].Result.Count().Should().Be(1);
+            commonResult[1].Result.First().N.Should().Be(4);
+
+            // It looks like the below test generates a wrong query. The server returns the following error:
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Tests/Linq/MongoQueryableTests.cs:1155]: As I wrote in the comment, it looks like the generated query is not correct. Should I create a new ticket for this error?

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs [hunk 2]
```diff
@@ -1276,5 +1304,57 @@ private bool TryTranslateSumResultOperator(PipelineExpression node, out BsonValu
...
+                if (startIndexOf != -1 && endIndexOf != -1 && endIndexOf > startIndexOf)
+                {
+                    endIndexOf = endIndexOf + MapTemplateEndDelimiter.Length;
+                    if (endIndexOf <= value.Length) return value.Substring(startIndexOf, endIndexOf - startIndexOf);
+                }
+
+                return null;
+            }
+
+            public static BsonValue ReplaceTemplates(BsonValue bsonValue, Dictionary<string,string> mapHistory)
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: During `Translation` processing in low-level expressions, we don't have information about which mapping ($map, $filter) should be done for parent nodes. This information will be available only at higher levels. So, in translation steps, we don't make any mappings. If it's necessary, we just mark these nodes as `template`s (which contain `key` values). 
In this method, these templates will be replaced on final query values, based on prepared _mapHistory.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs [hunk 3]
```diff
@@ -1276,5 +1304,57 @@ private bool TryTranslateSumResultOperator(PipelineExpression node, out BsonValu
...
         }
+
+        private static class TemplateHelper
+        {
+            private const string MapTemplateStartDelimiter = "#{{";
+            private const string MapTemplateEndDelimiter = "}}#";
+            private const string MapTemplate = MapTemplateStartDelimiter + "{{{0}}}" + MapTemplateEndDelimiter;
+
+            private static string GetNotReplacedTemplate(string value)
+            {
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs:1324]: Get first not replaced template

File: src/MongoDB.Driver/Linq/Expressions/IExpressionMemberInfo.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+namespace MongoDB.Driver.Linq.Expressions
+{
+    internal interface IExpressionMemberInfo
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/Linq/Expressions/IExpressionMemberInfo.cs]: This interface was added with idea that this logic can be added also to `DocumentExpression`. If it not so, It can be deleted.

File: src/MongoDB.Driver/Linq/Expressions/FieldExpression.cs [hunk 5]
```diff
@@ -60,6 +61,8 @@ public string FieldName
             get { return _fieldName; }
         }
 
+        public string OutOfCurrentScopePrefix { get; set; }
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Expressions/FieldExpression.cs]: Don't mix automatic properties with backing fields in the same class.

Add a _outOfCurrentScopePrefix field to the class.

My recommendation is don't use automatic properties for anything but the simplest classes (presumably data holding classes that have no logic).

File: src/MongoDB.Driver/Linq/Processors/SerializationBinder.cs [hunk 6]
```diff
@@ -151,6 +160,10 @@ protected override Expression VisitMember(MemberExpression node)
                                 mex);
                         }
                     }
+
+                    var rawParameterExpression = (node.Expression as ParameterExpression);
+                    if (_isOutOfCurrentScope && rawParameterExpression != null)
+                        SaveOutOfScopePrefix(newNode, rawParameterExpression.Name);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Processors/SerializationBinder.cs]: Always use curly braces around the body of the then clause even if it's only one line.

File: src/MongoDB.Driver/Linq/Expressions/IExpressionMemberInfo.cs (hunk 4 above)
Comment (rstam) [reply src/MongoDB.Driver/Linq/Expressions/IExpressionMemberInfo.cs]: This name confused me because I thought it was related to a LINQ Expression class or a reflection MethodInfo class, but it is neither.

How about something more direct like:

    IHasOutOfCurrentScopePrefix

File: src/MongoDB.Driver/Linq/Expressions/IExpressionMemberInfo.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        string OutOfCurrentScopePrefix { get; set; }
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Expressions/IExpressionMemberInfo.cs:20]: Looking at where and how this is set it seems like it would never have a dotted value. Would there not be some situations where a dotted field should be supported?

File: src/MongoDB.Driver/Linq/Processors/SerializationBinder.cs [hunk 8]
```diff
@@ -151,6 +160,10 @@ protected override Expression VisitMember(MemberExpression node)
                                 mex);
                         }
                     }
+
+                    var rawParameterExpression = (node.Expression as ParameterExpression);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Processors/SerializationBinder.cs:164]: Why "rawParameterExpression" instead of "parameterExpression".

What is "raw" about it?

File: src/MongoDB.Driver/Linq/Processors/SerializationBinder.cs [hunk 9]
```diff
@@ -397,5 +410,14 @@ private Expression BindClientSideProjector(MethodCallExpression node)
 
             return node;
         }
+
+        private void SaveOutOfScopePrefix(Expression expression, string value)
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Processors/SerializationBinder.cs]: Would a more accurate name be "SetOutOfCurrentScopePrefixIfPossible"?

Or do we even need this method? It's only called from one place.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs [hunk 10]
```diff
@@ -1276,5 +1303,57 @@ private bool TryTranslateSumResultOperator(PipelineExpression node, out BsonValu
             result = null;
             return false;
         }
+
+        private static class TemplateHelper
+        {
+            private const string MapTemplateStartDelimiter = "#{{";
+            private const string MapTemplateEndDelimiter = "}}#";
+            private const string MapTemplate = MapTemplateStartDelimiter + "{{{0}}}" + MapTemplateEndDelimiter;
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: The amount of escaping going on regarding the braces is confusing.

Should templates use a different character?

Comment (rstam) [reply src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: Is string based templating more appropriate than having a class representing a template? I haven't though it through... but string based templating sounds dangerous. What if a document just happens to have a field name that "looks" like a template to our code?

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs [hunk 11]
```diff
@@ -34,15 +33,24 @@ internal sealed class AggregateLanguageTranslator
...
+            if (builder._isTemplateMappingRequired)
+            {
+                result = TemplateHelper.ReplaceTemplates(result, builder._mapHistory);
+            }
+
+            return result;
         }
 
         private readonly AggregateStringTranslationMode _stringTranslationMode;
+        private readonly Dictionary<string, string> _mapHistory;
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: Not sure what you mean by "history".

Do you just mean "_templateMappings"?

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs [hunk 12]
```diff
@@ -590,12 +598,26 @@ private BsonValue TranslateSelect(SelectExpression node)
...
+                    return inputValue.ToString() + inValue.ToString().Substring(node.ItemName.Length + 2);
+                }
+                else
+                {
+                    _isTemplateMappingRequired = true;
+                    // If the field expression uses expression parameters outside of the current expression level 
+                    // then the generated value should be trimmed and do not have the expression prefix from the current level.
+                    // Here we just add a template which will be replaced in the end of transformation.
+                    return TemplateHelper.CreateTemplate(memberInfo.OutOfCurrentScopePrefix) + inValue.ToString().Substring(node.ItemName.Length + 2);
+                }
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs:623]: Should this be refactored? There is some repetition between the then and else clauses.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs [hunk 13]
```diff
@@ -741,10 +766,11 @@ private bool TryTranslateAnyResultOperator(PipelineExpression node, out BsonValu
                 else
                 {
                     var inValue = TranslateValue(FieldNamePrefixer.Prefix(whereExpression.Predicate, "$" + whereExpression.ItemName));
-
+                    var input = TranslateValue(whereExpression.Source);
+                    _mapHistory.Add(input.ToString(), whereExpression.ItemName);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: We have no idea what type of BsonValue input might be... why do we think converting it to a string using ToString() will make a suitable key for the map?

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs [hunk 14]
```diff
@@ -590,12 +598,26 @@ private BsonValue TranslateSelect(SelectExpression node)
...
+                {
+                    _isTemplateMappingRequired = true;
+                    // If the field expression uses expression parameters outside of the current expression level 
+                    // then the generated value should be trimmed and do not have the expression prefix from the current level.
+                    // Here we just add a template which will be replaced in the end of transformation.
+                    return TemplateHelper.CreateTemplate(memberInfo.OutOfCurrentScopePrefix) + inValue.ToString().Substring(node.ItemName.Length + 2);
+                }
             }
 
+            _mapHistory.Add(inputValue.ToString(), node.ItemName);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: I assume node.ItemName is a string representing the name of the parameter to the original Select expression.

Isn't using strings dangerous? What if there are two parameters of the same name at different levels.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs [hunk 15]
```diff
@@ -648,6 +670,7 @@ private BsonValue TranslateWhere(WhereExpression node)
             var inputValue = TranslateValue(node.Source);
             var condValue = TranslateValue(FieldNamePrefixer.Prefix(node.Predicate, "$" + node.ItemName));
 
+            _mapHistory.Add(inputValue.ToString(), node.ItemName);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: I don't understand this line.

inputValue is an arbitrary BsonValue whose ToString representation could be anything. Why would we expect that arbitrary string to be mappable to an ItemName (which I assume is actually a parameter name).

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs [hunk 16]
```diff
@@ -1276,5 +1303,57 @@ private bool TryTranslateSumResultOperator(PipelineExpression node, out BsonValu
...
+        private static class TemplateHelper
+        {
+            private const string MapTemplateStartDelimiter = "#{{";
+            private const string MapTemplateEndDelimiter = "}}#";
+            private const string MapTemplate = MapTemplateStartDelimiter + "{{{0}}}" + MapTemplateEndDelimiter;
+
+            private static string GetNotReplacedTemplate(string value)
+            {
+                var startIndexOf = value.IndexOf(MapTemplateStartDelimiter, StringComparison.Ordinal);
+                var endIndexOf = value.IndexOf(MapTemplateEndDelimiter, StringComparison.Ordinal);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: Are we assuming that there can only be one template (no nested templates)?

Otherwise we might need to search for the end index from the back.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs [hunk 17]
```diff
@@ -1276,5 +1303,57 @@ private bool TryTranslateSumResultOperator(PipelineExpression node, out BsonValu
...
+
+                return null;
+            }
+
+            public static BsonValue ReplaceTemplates(BsonValue bsonValue, Dictionary<string,string> mapHistory)
+            {
+                var value = bsonValue.ToString();
+                foreach (var map in mapHistory)
+                {
+                    value = value.Replace(string.Format(MapTemplate, map.Value), map.Key);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: It's really hard to figure out what string.Format is doing here. Why even use string.Format at all? Why not just use something like:

    var oldValue = $"#<{map.Value}>#"; // assuming using "<" in templates instead of "{" which requires escaping
    var newValue = map.Key;
    value = value.Replace(oldValue, newValue);

For that matter, why doesn't map.Value already have the finished template already? Why do we have to do any formatting here?

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs [hunk 18]
```diff
@@ -1276,5 +1303,57 @@ private bool TryTranslateSumResultOperator(PipelineExpression node, out BsonValu
...
+                    // We need to handle this case additionally.
+                    value = value.Replace(topLevelParameter, mapHistory.Last().Key);
+                }
+
+                if (value.Contains(MapTemplateStartDelimiter) || value.Contains(MapTemplateEndDelimiter))
+                {
+                    throw new MongoException("The query transformation has not been finished: Incorrect parameters mapping.");
+                }
+
+                return BsonDocument.Parse(value);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: How do you know value is a parseable BsonDocument? It was only a BsonValue coming in.

File: tests/MongoDB.Driver.Tests/Linq/MongoQueryableTests.cs [hunk 19]
```diff
@@ -1136,6 +1134,35 @@ public void Select_followed_by_group()
                 "{ $group: { _id: '$First', Stuff: { $push: { Id: '$Id', Second: '$Second' } } } }");
         }
 
+
+        [SkippableFact]
+        public void Select_with_using_top_level_expression_input_parameters_inside_children_levels()
+        {
+            RequireServer.Check().Supports(Feature.AggregateArrayFilter);
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/MongoQueryableTests.cs]: I haven't been able to figure how the AggregateArrayFilter feature is required for this test.

Comment (rstam) [reply tests/MongoDB.Driver.Tests/Linq/MongoQueryableTests.cs]: I see my confusion, there are two very similarly named features which represent completely different things:

    Feature.AggregateArrayFilter
    Feature.ArrayFilters

That's a bummer.

File: tests/MongoDB.Driver.Tests/Linq/IntegrationTestBase.cs [hunk 20]
```diff
@@ -81,7 +81,9 @@ private void InsertFirst()
                                     {
                                         D = "Delilah"
                                     }
-                            }
+                            },
+                            N = 1,
+                            Y = new List<E> { new E { F = 101 }, new V { W = 102 } }
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/IntegrationTestBase.cs]: This class is getting out of control. Too many different tests depend on it and that's why it has way too many fields.

File: src/MongoDB.Driver/Linq/Expressions/IExpressionMemberInfo.cs (hunk 4 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Expressions/IExpressionMemberInfo.cs]: It can be renamed.

File: src/MongoDB.Driver/Linq/Expressions/FieldExpression.cs (hunk 5 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Expressions/FieldExpression.cs]: Ok

File: src/MongoDB.Driver/Linq/Processors/SerializationBinder.cs (hunk 6 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Processors/SerializationBinder.cs]: Ok

File: src/MongoDB.Driver/Linq/Expressions/IExpressionMemberInfo.cs (hunk 7 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Expressions/IExpressionMemberInfo.cs:20]: In the current implementation, this field has value only from the lowest expression (document itself) => so yeah, we can have only field values without `dot`. If it is necessary we can update this logic, but now - it's not needed.

File: src/MongoDB.Driver/Linq/Processors/SerializationBinder.cs (hunk 8 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Processors/SerializationBinder.cs:164]: "raw" - in this context, it means not visited yet (without any transformations). Maybe it's already clear from the code. I can rename it.

File: src/MongoDB.Driver/Linq/Processors/SerializationBinder.cs (hunk 9 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Processors/SerializationBinder.cs]: I took into account that this logic could be called not only for lowest expression field level  (for fields with `dot` too). In this case, it will be simply to call this method for them too.
Renaming looks good. I will do it.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs (hunk 11 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: `history` means - just a collection of `maps` which were introduced in a query (`$map` and `$filter`).
`"_templateMappings"` - Ok, I will rename.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs (hunk 12 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs:623]: I decided that this way can be more clear for other people. But no problem, I will do it.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs (hunk 14 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: I will check it closely. 

>Isn't using strings dangerous?

I guess it's safe, since it's already string.

>I assume node.ItemName is a string representing the name of the parameter to the original Select expression.

at least not always. 

        g.G.Min(g2 => g2.N)

the `Binder` logic generates the following code for above `Linq` expression.

        Min({document}{G}.Select({document}{N))

So, this code can be generated not only by Linq `Select`.

I also check the `inputValue`. I didn't expect that this value can be `BsonArray`. I will check how my logic will work for this(and similar) cases.

I will come back with the clarifications.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs (hunk 16 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: >Are we assuming that there can only be one template (no nested templates)?

yes, for the current implementation.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs (hunk 18 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: Agree. I will replace this on `BsonValue.Create(value)`.

File: tests/MongoDB.Driver.Tests/Linq/MongoQueryableTests.cs (hunk 19 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Linq/MongoQueryableTests.cs]: Maybe I'm wrong.
When I launched this build on evergreen, I caught the following error:
`MongoDB.Driver.MongoCommandException : Command aggregate failed: exception: invalid operator '$filter'.`
Since this filter has been used inside `aggregate` command:

        aggregate([{ "$project" : { "Result" : { "$filter" : { "input" : "$G", "as" : "g1", "cond" ...

, I've decided that it's `Aggregate filter`. 
As I see from this link, `arrayFilter` it's a little bit another thing: https://docs.mongodb.com/manual/reference/operator/update/positional-filtered/. Could you please clarify this point? :)

File: tests/MongoDB.Driver.Tests/Linq/IntegrationTestBase.cs (hunk 20 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Linq/IntegrationTestBase.cs]: Completely agree. Maybe we can refactor it by class generic param + common(somehow) logic for collection initialization?

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs (hunk 10 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: >What if a document just happens to have a field name that "looks" like a template to our code?

I expect that document cannot have a field name with `#{` or '}#' symbols since `AggregateLanguageTranslator` receives a c# linq expression which doesn't allow these field names.

>than having a class representing a template?

I'm not sure what you are suggesting. Do you mean creation some json template (with class representation) which will be added to translated value instead of `MapTemplate`. 
Having something like this:
        
        JsonTemplate.Create(memberInfo.OutOfCurrentScopePrefix) + inValue.ToString().Substring(...

Instead of:
        
        TemplateHelper.CreateTemplate(memberInfo.OutOfCurrentScopePrefix) + inValue.ToString().Substring(...

If so, how do you suggest to replace template values after translation?

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs (hunk 13 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: I need to check once again this case. AS I wrote in some other point:
>I didn't expect that this value can be BsonArray. I will check how my logic will work for this(and similar) cases.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs (hunk 15 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: I wrote in other places:
>I didn't expect that this value can be BsonArray. I will check how my logic will work for this(and similar) cases.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs (hunk 17 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: Agree, it can be replaced on your suggestion.

File: src/MongoDB.Driver/Linq/Expressions/FieldExpression.cs (hunk 5 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Expressions/FieldExpression.cs]: Just to clarify, what's the reason for that? If it's about having the same using way for all properties, I would say that a better way just conversion properties with backing fields to autoproperty

File: src/MongoDB.Driver/Linq/Expressions/IExpressionMemberInfo.cs (hunk 4 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Expressions/IExpressionMemberInfo.cs]: Done

File: src/MongoDB.Driver/Linq/Processors/SerializationBinder.cs (hunk 6 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Processors/SerializationBinder.cs]: Done

File: src/MongoDB.Driver/Linq/Processors/SerializationBinder.cs (hunk 9 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Processors/SerializationBinder.cs]: Done

File: src/MongoDB.Driver/Linq/Expressions/FieldExpression.cs (hunk 5 above)
Comment (rstam) [reply src/MongoDB.Driver/Linq/Expressions/FieldExpression.cs]: The reasons are:

1. It is helpful if all the members that hold state are listed together so that when reading the class one can understand what state an object holds by looking at one place
2. Any time you add any validation code to a property setter (null checks etc..) you need a backing field
3. You often need fields that aren't necessarily exposed publicly via properties

Since one is likely to end up with backing fields anyway for all but the simplest classes let's just always use them.

In the case of read-only properties manually adding an explicit backing field is one line of code.

Definitely let's never mix backing fields with automatic properties because then it's very easy to misinterpret what state an object can hold.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs (hunk 10 above)
Comment (rstam) [reply src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: I don't actually have an alternate suggestion. But I am expressing the feedback that specially formatted strings feel like a potentially dangerous hack.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs (hunk 18 above)
Comment (rstam) [reply src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: BsonValue.Create doesn't sound right either. Since value is a string BsonValue.Create will just return a BsonString.

File: tests/MongoDB.Driver.Tests/Linq/MongoQueryableTests.cs (hunk 19 above)
Comment (rstam) [reply tests/MongoDB.Driver.Tests/Linq/MongoQueryableTests.cs]: My point was simply that we have two Features with very similar names and I was confused and mixed them up. At first I didn't understand why you added AggregateArrayFilters when we already had ArrayFilters.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs [hunk 21]
```diff
@@ -34,17 +33,31 @@ internal sealed class AggregateLanguageTranslator
...
 
         private AggregateLanguageTranslator(ExpressionTranslationOptions translationOptions)
         {
             translationOptions = translationOptions ?? ExpressionTranslationOptions.Default;
             _stringTranslationMode = translationOptions.StringTranslationMode.GetValueOrDefault(AggregateStringTranslationMode.Bytes);
         }
 
+        private BsonValue Translate(Expression node)
+        {
+            _templatesState = TemplatesState.Create();
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs:51]: Creation the object which will accumulate information about templates.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs [hunk 22]
```diff
@@ -34,17 +33,31 @@ internal sealed class AggregateLanguageTranslator
...
+            var result = TranslateValue(node);
+            if (_templatesState.IsMappingRequired)
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs:53]: if we haven't added any template => Mapping is not required

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs [hunk 23]
```diff
@@ -34,17 +33,31 @@ internal sealed class AggregateLanguageTranslator
...
+            {
+                result = _templatesState.ReplaceTemplates(result);
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs:55]: Start template replacing

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs [hunk 24]
```diff
@@ -596,6 +611,7 @@ private BsonValue TranslateSelect(SelectExpression node)
                 return inputValue.ToString() + inValue.ToString().Substring(node.ItemName.Length + 2);
             }
 
+            _templatesState.AddMapping(node.ItemName);
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs:626]: This saving is required to understand whether the out of scope parameter is defined in the top level or not. If there is mapping - the out of scope parameter was defined inside a not top-level expression

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs [hunk 25]
```diff
@@ -1276,5 +1297,203 @@ private bool TryTranslateSumResultOperator(PipelineExpression node, out BsonValu
             result = null;
             return false;
         }
+
+        internal class TemplatesState
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs:1313]: I guess this class requires a number of unit tests. I will add them if the current solution is ok.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs [hunk 26]
```diff
@@ -1276,5 +1297,203 @@ private bool TryTranslateSumResultOperator(PipelineExpression node, out BsonValu
...
+                    var dotIndex = value.IndexOf('.');
+                    if (dotIndex != -1)
+                    {
+                        return $"{value.Substring(dotIndex + 1)}";
+                    }
+                }
+                return value;
+            }
+
+            private static Dictionary<string, List<string>> CollectTemplates(string value)
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: In result:
Key - is an out of scope input expression parameter.
Value - is a collection with all fields which are based on the key.

File: src/MongoDB.Driver/Linq/Expressions/FieldExpression.cs (hunk 5 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Expressions/FieldExpression.cs]: Ok

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs (hunk 13 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: I've changed this logic.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs (hunk 18 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: I've changed this code.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs [hunk 27]
```diff
@@ -1276,5 +1297,203 @@ private bool TryTranslateSumResultOperator(PipelineExpression node, out BsonValu
...
+                        {
+                            throw new MongoException("The query transformation has not been finished: Incorrect template format.");
+                        }
+                    }
+                }
+
+                return templates;
+            }
+
+            private static string CreateDocument(string ket, string value)
```
Comment (craiggwilson) [inline src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: key?

File: tests/MongoDB.Driver.Tests/Linq/MongoQueryableTests.cs (hunk 1 above)
Comment (craiggwilson) [reply tests/MongoDB.Driver.Tests/Linq/MongoQueryableTests.cs:1155]: that means that the `$map` is returning null, which is weird. The only way that should happen is if `G.B.S` doesn't exist in some document. If that's true, then we need to add in null checks, which really suck to have. But yes, probably file a new bug ticket.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs (hunk 14 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: Does this test help: Select_with_using_several_expression_parameters_from_different_levels_inside_children_level
?

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs (hunk 25 above)
Comment (craiggwilson) [reply src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs:1313]: Yeah... the unit tests will show me what is actually going on in here. I think I get it, but they will clarify.

File: tests/MongoDB.Driver.Tests/Linq/MongoQueryableTests.cs [hunk 28]
```diff
@@ -1136,6 +1134,158 @@ public void Select_followed_by_group()
...
+            commonResult[0].Result.Count().Should().Be(1);
+            commonResult[0].Result.First().Should().Be(1);
+            commonResult[1].Result.Count().Should().Be(1);
+            commonResult[1].Result.First().Should().Be(3);
+        }
+
+        //[SkippableFact]
+        //public void Select_with_using_top_level_expression_input_parameters_inside_children_levels_and_with_inner_select22()
+        //{
+        //this test doesn't return correct Select result, but it looks like it's minor problem.
```
Comment (craiggwilson) [inline tests/MongoDB.Driver.Tests/Linq/MongoQueryableTests.cs]: Is this in scope to solve here? The title of 2269 seems to indicate it is...

**Review by craiggwilson — COMMENTED** [review]
This is a clever solution. The mapping system is what we do in the BI-Connector, although for a slightly different reason. Just a few comments.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs (hunk 27 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: Sorry, I've fixed this. Just pushed not last version :(

Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs]: Last version has been already pushed.

File: tests/MongoDB.Driver.Tests/Linq/MongoQueryableTests.cs (hunk 28 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Linq/MongoQueryableTests.cs]: Good point. I wasn't sure that we should care about this case in the scope of this ticket.
This test is about having an out of scope expression parameter in `Select` predicate which doesn't look like often situation. But I've fixed this case too. See my last commit.

I also had some other tests which were not included to the final commit since they had Linq expressions which are not so close to this ticket description. I guess I will create tests for them and then we will decide should we worry about them now or not. Maybe I will create new tickets for them.

File: tests/MongoDB.Driver.Tests/Linq/MongoQueryableTests.cs (hunk 1 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Linq/MongoQueryableTests.cs:1155]: >But yes, probably file a new bug ticket.

@rstam do you mind? I can create this ticket.

File: src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs (hunk 25 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs:1313]: I will add it tomorrow. I will let you know.

**Comment by DmitryLukyanov** [comment]
Closed since there is a better solution which requires a big refactoring
